File: src/main/scala/e2m/mongodb/MongoExport.scala
package e2m

import com.mongodb.client.result.InsertOneResult
import com.mongodb.client.{MongoClient, MongoClients, MongoCollection, MongoDatabase}
import org.bson.types.ObjectId
import org.bson.{BsonObjectId, BsonString, BsonValue, Document}

import java.util
import scala.util.{Failure, Success}
import scala.util.Try
import scala.jdk.CollectionConverters.*

class MongoExport(database: String,
                  collection: String,
                  clear: Boolean = false,
                  host: Option[String] = None,
                  port: Option[Int] = None,
                  user: Option[String] = None,
                  password: Option[String] = None) {
  require ((user.isEmpty && password.isEmpty) || (user.nonEmpty && password.nonEmpty))

  private val hostStr: String = host.getOrElse("localhost")
  private val portStr: String = port.getOrElse(27017).toString
  private val usrPswStr: String = user match
    case Some(usr) => s"$usr:${password.get}@"
    case None => ""

  // mongodb://[username:password@]host[:port][,...hostN[:port]][/[database][?parameter_list]]
  private val mongoUri: String = s"mongodb://$usrPswStr$hostStr:$portStr"
  private val mongoClient: MongoClient = MongoClients.create(mongoUri)
  private val dbase: MongoDatabase = mongoClient.getDatabase(database)

  private val coll: MongoCollection[Document] =
    if (clear)
      dbase.getCollection(collection).drop()
      dbase.getCollection(collection)
    else dbase.getCollection(collection)

  def close(): Try[Unit] = Try(mongoClient.close())

  def insertDocument(doc: String): Try[String] = {
    Try {
      val document: Document = Document.parse(doc)
      //coll.insertOne(document).getInsertedId.asString().toString
      coll.insertOne(document).getInsertedId.asObjectId().toString
    }
  }

  def insertDocuments(docs: Seq[String]): Try[Seq[String]] = {
    Try {
      val documents: util.List[Document] = docs.map(Document.parse).asJava
      val insertedIds: util.Collection[BsonValue] = coll.insertMany(documents).getInsertedIds.values()

      insertedIds.asScala.toSeq.map(_.asObjectId().toString)
    }
  }
}

File: python/e2m/mongodb/mongoExport.py
import pymongo


class MongoExport:

    def __init__(self, database, collection, clear=False, host=None, port=None, user=None, password=None):
        self.__host = host or "localhost"
        self.__port = port or 27019
        self.__user = user
        self.__password = password
        self.__usrPsrStr = "" if not user and not password else f"{user}:{password}@"
        self.__mongoUri = f"mongodb://{self.__usrPsrStr}{self.__host}:{self.__port}"
        self.__mongoClient = pymongo.MongoClient(self.__mongoUri)
        self.__dbase = self.__mongoClient.get_database(database)
        self.__clear = self.__dbase.get_collection(collection) if not clear else {
            self.drop_coll(self.__dbase.get_collection(collection))
        }
        self.__coll = self.__dbase.get_collection(collection)

    def insert(self, document):
        self.__coll.insert_one(document)

    def close(self):
        self.__mongoClient.close()

    @staticmethod
    def drop_coll(collection):
        collection.drop()
        return collection
